test/jit/test_hooks.py

```diff
@@ -0,0 +1,357 @@
+            RuntimeError,
+            "Pre-hook '.+' on .+ has at least two different python "
+            "definitions. Please use unique names for all hooks.",
+        ):
+            torch.jit.script(m)
+
+        m = ModuleForwardSingleInput("outer_mod_name", "inner_mod_name")
+
+        def hook(self, input: Tuple[str], output: str):
+            return "This is the first hook"
+
+        m.submodule.register_forward_hook(hook)
+
+        def hook(self, input: Tuple[str]):
+            return "This is the second hook"
+
+        m.submodule.register_forward_hook(hook)
+
+        with self.assertRaisesRegex(
+            RuntimeError,
+            "Hook '.+' on .+ has at least two different python "
+            "definitions. Please use unique names for all hooks.",
+        ):
+            torch.jit.script(m)
+
+    def test_module_direct_forward_invocation(self):
+        # Test that hooks are only invoked when the module is
+        # called directly and not when forward is called.
+        m = ModuleForwardSingleInput("outer_mod_name", "inner_mod_name")
+
+        def pre_hook(self, input: Tuple[str]) -> Tuple[str]:
+            return ("pre_hook_override_name",)
+
+        def forward_hook(self, input: Tuple[str], output: str):
+            assert self.name == "outer_mod_name"
+            assert input == ("pre_hook_override_name",)
+            output = output + "_fh"
+            return output
+
+        m.register_forward_pre_hook(pre_hook)
+        m.register_forward_hook(forward_hook)
+
+        m_scripted = torch.jit.script(m)
+
+        self.assertEqual(m.forward("a"), m_scripted.forward("a"))
+        self.assertNotEqual(m_scripted("a"), m_scripted.forward("a"))
+
+    def test_submodule_direct_forward_invocation(self):
+        m_submod_forward_call = ModuleDirectFowardSubmodCall(
+            "outer_mod_name", "inner_mod_name"
+        )
+        m_submod_call = ModuleForwardSingleInput("outer_mod_name", "inner_mod_name")
+
+        def pre_hook(self, input: Tuple[str]) -> Tuple[str]:
+            return ("pre_hook_override_name",)
+
+        def forward_hook(self, input: Tuple[str], output: str):
+            assert input == ("pre_hook_override_name",)
+            return output + "_fh"
+
+        m_submod_forward_call.submodule.register_forward_pre_hook(pre_hook)
+        m_submod_forward_call.submodule.register_forward_hook(forward_hook)
+        m_submod_call.submodule.register_forward_pre_hook(pre_hook)
+        m_submod_call.submodule.register_forward_hook(forward_hook)
+
+        m_submod_forward_call_scripted = torch.jit.script(m_submod_forward_call)
+        m_submod_call_scripted = torch.jit.script(m_submod_call)
+
+        self.assertEqual(
+            m_submod_forward_call_scripted("a"), m_submod_forward_call("a")
+        )
+        self.assertNotEqual(
+            m_submod_forward_call_scripted("a"), m_submod_call_scripted("a")
+        )
+
+    """ TODO: add this test back once figured out how to print 
```

Comment:
Nit: I would prefer to use `@unittest.skip` here instead of commenting the test out like this but it's just personal preference.